vcr/issuer/issuer.go

```diff
@@ -0,0 +1,121 @@
+package issuer
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	ssi "github.com/nuts-foundation/go-did"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/crypto"
+	"github.com/nuts-foundation/nuts-node/vcr/credential"
+	"github.com/nuts-foundation/nuts-node/vcr/signature"
+	"github.com/nuts-foundation/nuts-node/vcr/signature/proof"
+	vdr "github.com/nuts-foundation/nuts-node/vdr/types"
+	"time"
+)
+
+// NewIssuer creates a new issuer which implements the Issuer interface.
+func NewIssuer(store Store, publisher Publisher, docResolver vdr.DocResolver, keyStore crypto.KeyStore) Issuer {
+	resolver := vdrKeyResolver{docResolver: docResolver, keyResolver: keyStore}
+	return &issuer{
+		store:       store,
+		Publisher:   publisher,
+		keyResolver: resolver,
+	}
+}
+
+type issuer struct {
+	store       Store
+	Publisher   Publisher
+	keyResolver keyResolver
+}
+
+// Issue creates a new credential, signs, stores it.
+// If publish is true, it publishes the credential to the network using the configured Publisher
+// Use the public flag to pass the visibility settings to the Publisher.
+func (i issuer) Issue(credentialOptions vc.VerifiableCredential, publish, public bool) (*vc.VerifiableCredential, error) {
+	createdVC, err := i.buildVC(credentialOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	validator, _ := credential.FindValidatorAndBuilder(*createdVC)
+	if err := validator.Validate(*createdVC); err != nil {
+		return nil, err
+	}
+
+	// TODO: Store credential in the store
+	if err = i.store.StoreCredential(*createdVC); err != nil {
+		return nil, fmt.Errorf("unable to store the issued credential: %w", err)
+	}
+
+	if publish {
+		if err := i.Publisher.PublishCredential(*createdVC, public); err != nil {
+			return nil, fmt.Errorf("unable to publish the issued credential: %w", err)
+		}
+	}
+	return createdVC, nil
+}
+
+func (i issuer) buildVC(credentialOptions vc.VerifiableCredential) (*vc.VerifiableCredential, error) {
+	if len(credentialOptions.Type) != 1 {
+		return nil, errors.New("can only issue credential with 1 type")
+	}
+
+	// find issuer
+	issuer, err := did.ParseDID(credentialOptions.Issuer.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse issuer: %w", err)
+	}
+
+	credentialID, _ := ssi.ParseURI(fmt.Sprintf("%s#%s", issuer.String(), uuid.New().String()))
+	unsignedCredential := vc.VerifiableCredential{
+		Context:           append(credentialOptions.Context, vc.VCContextV1URI()),
```

Comment:
maybe I missed it, but I thought the default value is the Nuts context?

Reply:
Not anymore. I want these api's to be more lowlevel. This allows for other context like schema.org or perhaps the KIK-V credentials.